### ISSUE TITLE
Mask extension to --max-length doesn't work in combination with wordlist

### ISSUE THREAD

**Issue by gwd** [openwall/john#4208]
The `--max-length` parameter doesn't seem to extend `--mask` values when used in combination with `--wordlist`.

### Steps to reproduce ###

The following command 

```
./john "--mask=base?a" --max-length=8 test.john
```

Will try `base` with 1, 2, 3, or 4 characters appended to it.  But the following command sequence:

```
echo "base" > test.lst
./john -w=test.lst "--mask=?w?a" --max-length=8 test.john
```

Will only try `base` with a single character appended to it.

### System configuration ###
Attach details about your OS and about JtR, including:
- `$ ./john --list=build-info`.

```
Version: 1.9.0-jumbo-1+bleeding-d41d53d95 2020-02-03 14:21:39 +0100
Build: linux-gnu 64-bit x86_64 AVX AC OMP
SIMD: AVX, interleaving: MD4:3 MD5:3 SHA1:1 SHA256:1 SHA512:1
CPU tests: AVX
$JOHN is ./
Format interface version: 14
Max. number of reported tunable costs: 4
Rec file version: REC4
Charset file version: CHR3
CHARSET_MIN: 1 (0x01)
CHARSET_MAX: 255 (0xff)
CHARSET_LENGTH: 24
SALT_HASH_SIZE: 1048576
SINGLE_IDX_MAX: 32768
SINGLE_BUF_MAX: 4294967295
Effective limit: Max. KPC 32768
Max. Markov mode level: 400
Max. Markov mode password length: 30
gcc version: 9.2.1
GNU libc version: 2.29 (loaded: 2.29)
Crypto library: OpenSSL
OpenSSL library version: 01010104f
OpenSSL 1.1.1d  10 Sep 2019
File locking: fcntl()
fseek(): fseek
ftell(): ftell
fopen(): fopen
memmem(): System's
times(2) sysconf(_SC_CLK_TCK) is 100
Using times(2) for timers, resolution 10 ms
HR timer claimed resolution 1 ns, observed 1 us
Total physical host memory: 369 GB
Available physical host memory: 366 GB
```

**Comment by magnumripper** [comment]
This works as intended, alas the documentation on stretching is "a bit sparse".

In hybrid mask mode, the mask is never stretched but the `--min-len` and `--max-len` options works just like they do in wordlist-only: If the ultimate candidate is out of bounds the word will be rejected.

**Comment by gwd** [comment]
OK, but suppose I know that the "modus operandi" is to have a base password + some suffix.

If I have only one 'base', but don't know how long the suffix is, I can search for it like this:

```
./john --mask="$BASE\a" --max-length=$LARGE
```

And JtR will start with $BASE and search for increasing lengths of suffixes.  If I have several "base" values, and know how long the suffix is, I can search for it like this:

```
../john -w=base.lst --mask="\w\a\a\a\a"
```

But suppose I have several "base" values, but *don't* know how long the suffix is.  Is there any way of doing that search, short of just manually calling `./john --mask=$BASE\a` on each different $BASE manually?

IOW -- if this isn't a bug, it's a feature request. :-)

**Comment by magnumripper** [comment]
I see what you mean but it's not currently possible (without running several sessions). This will be solved with #2310 and/or #4016.

**Comment by solardiz** [comment]
It's currently possible to use wordlist rules (with preprocessor) instead of mask, e.g.:

```
[List.Rules:AddMasks]
Az"[a-z]"
Az"[a-z][a-z]"
Az"[a-z][a-z][a-z]"
Az"[a-z][a-z][a-z][a-z]"
```

```
./john -w=base.lst --rules=addmasks hash
```

This way, only one session needs to be run. Drawbacks: have to list all "masks", slow startup if these expand to a lot of rules (we validate syntax of expanded rules at startup).
